lib/grape/extensions/hash.rb

```diff
@@ -12,8 +12,8 @@ module ParamBuilder
 
         def build_params
           rack_params.deep_dup.tap do |params|
-            params.deep_merge!(grape_routing_args) if env.key?(Grape::Env::GRAPE_ROUTING_ARGS)
             params.deep_symbolize_keys!
+            params.deep_merge!(grape_routing_args.deep_symbolize_keys!) if env.key?(Grape::Env::GRAPE_ROUTING_ARGS)
```

Comment:
This modifies the input `grape_routing_args.deep_symbolize_keys!` which may be an undesirable side-effect. 

Either remove `!` which would make a copy, or perform `grape_routing_args.deep_symbolize_keys!` elsewhere in all conditions ensuring that's intended?

Reply:
Let's break this up to avoid hiding a side effect:

```ruby
if env.key?(Grape::Env::GRAPE_ROUTING_ARGS)
   grape_routing_args.deep_symbolize_keys!
   params.deep_merge!(grape_routing_args)
end
```